### PR TITLE
Add new option to vibrate when quick settings tile is clicked

### PR DESCRIPTION
**Pull request by dshokouhi**
## Summary

Fixes: #3119 by introducing a new per tile setting (off by default) so the user can turn on vibration. As soon as the tile is clicked we will vibrate, then if the call fails we will vibrate again for longer.

## Screenshots

image

## Link to pull request in Documentation repository

Documentation: home-assistant/companion.home-assistant#868

## Any other notes

Leaving as a draft until after #3121  to avoid a conflict. Submitting early to get feedback.

### CODE REVIEW

**Comment by icegothic** [comment]
If I may, is'nt 500ms and 1000ms a little long for vibrations? When I suggested this feature, I was thinking more of a haptic feedback to confirm that I tapped the tile properly. When I used Tasker&QuickTask to manage my tiles, I was using a 5ms to 20ms vibration.

**Comment by dshokouhi** [comment]
> If I may, is'nt 500ms and 1000ms a little long for vibrations?

This matches our haptic feedback durations in our WebView Activity for success and failure.



> When I suggested this feature, I was thinking more of a haptic feedback to confirm that I tapped the tile properly

I would love to use the actual success and failure haptic feedback constants but those are only available inside an activity so we cannot utilize them in quick settings so to make it compatible with all devices we go with a set duration.

**Comment by icegothic** [comment]
Is there a way to make the duration customizable by the user (in ms)?

**Comment by dshokouhi** [comment]
> Is there a way to make the duration customizable by the user (in ms)?

I want to avoid adding so many options and if we do that here then it will be expected to do it in our WebView Activity.

**Comment by icegothic** [comment]
I get that :-) but half a second vibration for every tile tap is way to long for me, I'll end up not using the feature I believe :-(

**Comment by dshokouhi** [comment]
> I get that :-) but half a second vibration for every tile tap is way to long for me, I'll end up not using the feature I believe :-(

if we can find the appropriate vibration length for the haptic constants then it would make sense to update both areas. Its important for the app to be consistent in places. Every user is going ot have their own idea what vibration is too much or too little so it won't be easy to please everyone.

**Comment by icegothic** [comment]
Allright, I'll wait for the release and see what it looks like :-) thanx for your consideration.

**Comment by dshokouhi** [comment]
@icegothic I made a change that should make this more suitable for your use case and others. If the device supports it (which yours does) we will now use a vibration effect that indicates a click. On older devices we will continue to vibrate for longer.

File: app/src/main/java/io/homeassistant/companion/android/settings/qs/views/ManageTilesView.kt [hunk 1]
```diff
@@ -154,6 +155,17 @@ fun ManageTilesView(
                         }
                     }
                 }
+
+                Row(verticalAlignment = Alignment.CenterVertically) {
+                    Text(
+                        text = stringResource(R.string.tile_vibrate),
+                        fontSize = 15.sp
+                    )
+                    Checkbox(
```
Comment (jpelgrom) [inline app/src/main/java/io/homeassistant/companion/android/settings/qs/views/ManageTilesView.kt:164]: Use a `Switch`, checkboxes are for selecting [one or more in a list](https://m3.material.io/components/checkbox/overview)

File: common/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -955,4 +955,5 @@
     <string name="sensor_description_daily_calories">The total number of calories over a day (including both BMR and active calories), where the previous day ends and a new day begins at 12:00 AM local time.</string>
     <string name="sensor_name_daily_steps">Daily Steps</string>
     <string name="sensor_description_daily_steps">The total step count over a day, where the previous day ends and a new day begins at 12:00 AM local time.</string>
+    <string name="tile_vibrate">Should vibrate when tile is clicked?</string>
```
Comment (jpelgrom) [inline common/src/main/res/values/strings.xml:958]: Why is this a question and not simply "Vibrate when clicked"?

Comment (jpelgrom) [reply common/src/main/res/values/strings.xml]: It is still a question 😉

Other on/off settings don't have a question mark

File: app/src/main/java/io/homeassistant/companion/android/settings/qs/views/ManageTilesView.kt [hunk 3]
```diff
@@ -154,6 +155,17 @@ fun ManageTilesView(
...
+
+                Row(verticalAlignment = Alignment.CenterVertically) {
+                    Text(
+                        text = stringResource(R.string.tile_vibrate),
+                        fontSize = 15.sp
+                    )
+                    Switch(
+                        checked = viewModel.selectedShouldVibrate,
+                        onCheckedChange = { viewModel.selectedShouldVibrate = it }
+                    )
```
Comment (jpelgrom) [inline app/src/main/java/io/homeassistant/companion/android/settings/qs/views/ManageTilesView.kt:167]: The switch's "thumb" lacks contrast when off and using dark mode:

Setting 'Vibrate when clicked' with a switch, where the switch "thumb"/circle blends into the background

This is solved everywhere else by specifying a different color like this:
```kotlin
colors = SwitchDefaults.colors(uncheckedThumbColor = colorResource(commonR.color.colorSwitchUncheckedThumb))
```

(In light mode it is easier to see due to the shadow)

**Review by jpelgrom — COMMENTED** [review]
I think the aim was to use the same vibration as clicks in the frontend, but this one feels different on my device (Pixel 4a): much shorter and less buzzing (it's an improvement IMO!).

File: app/src/main/java/io/homeassistant/companion/android/settings/qs/views/ManageTilesView.kt (hunk 3 above)
Comment (dshokouhi) [reply app/src/main/java/io/homeassistant/companion/android/settings/qs/views/ManageTilesView.kt:169]: Thank you that does look better!

**Review by jpelgrom — APPROVED** [review]

**Review by JBassett — APPROVED** [review]

**Pull request merged**